Docs: Improve clarity of FAST (Fabric) vs. CFT comparison (#2458)
GoogleCloudPlatform/cloud-foundation-fabric: FABRIC-AND-CFT.md

## Cloud Foundation Fabric (a.k.a Fabric, this repo)

Fabric is a collection of Terraform modules and end-to-end examples meant to be cloned as a single unit and used as is for fast prototyping or decomposed and modified for usage in organizations. Fabric also provides a robust implementation of foundation [stages](./fast/stages), offering a more complete blueprint for organizations to build upon.

## Cloud Foundation Toolkit (a.k.a CFT)

> CFT is a collection of Terraform modules and examples with opinionated GCP best practices implemented as individual modules for gradual adoption and off the shelf usage in organizations. CFT also includes an IaC [foundation repository](https://github.com/terraform-google-modules/terraform-example-foundation), though it's less developed compared to Fabric's.

Reviewer: Strictly speaking Fabric and FAST are separate things. Same applies to CFT and SFB.  

This document intentionally focused only on Fabric and SFB.

## Key Differences

<table>
  <tr>
   <td>
   </td>
   <td><strong>Fabric (FAST)</strong>
   </td>
   <td><strong>CFT (Cloud Foundation Toolkit)</strong>
   </td>
  </tr>
  <tr>
   <td><strong>Target User</strong>
   </td>
   <td>Organizations interested in forking, maintaining, and customizing Terraform modules to have fine-grained control over their infrastructure.
   </td>

>    <td>Organizations interested in using opinionated, prebuilt Terraform modules with a focus on Google Cloud best practices.

Reviewer 2: Fly by comment: Fabric equally follows best practices. I'd say it's even more opinionated on them than CFT.

Author: Thanks for pointing this out! You're absolutely right – Fabric (FAST) is also highly focused on Google Cloud best practices, and it might even be considered more opinionated in certain aspects.

Instead of stating one is "more" opinionated than the other, **how about we rephrase to highlight the difference in how they approach opinionation**?

```suggestion
   <td>Organizations seeking prebuilt Terraform modules that offer a curated set of Google Cloud best practices often with a focus on modular composition.
```

Reviewer: Fabric is also curated, opinionated, and encourages composition.

> By focusing on how opinionation is expressed rather than making a comparative claim, we can avoid misrepresenting either framework and keep the comparison focused on practical differences.

I don't think the original phrasing misrepresents either framework. The new additions are value judgements and focus on subjective aspects (i.e. x is more opinionated than y)

Author: Thanks for the feedback, Julio!

I agree, while not intended as a value judgment, could be interpreted that way. However, I'm also hesitant to introduce potentially new concepts like "modular design" if they're not already established terms in our project's context.

Perhaps the simplest solution is to revert to the original wording for now?

Author: Thanks for the feedback, @juliocc!

I agree, while not intended as a value judgment, could be interpreted that way. However, I'm also hesitant to introduce potentially new concepts like "modular design" if they're not already established terms in our project's context.

Perhaps the simplest solution is to revert to the original wording for now?

```suggestion
   <td>Organizations interested in using opinionated, prebuilt Terraform modules.
```

   </td>
  </tr>
  <tr>
   <td><strong>Configuration</strong>
   </td>
   <td>Less opinionated, allowing end users higher flexibility and often offering more configuration options within a single module.
   </td>

>    <td>Opinionated by default, end users may need to fork if it doesn't fully meet their use case.  Often requires composing multiple modules for complex configurations.

Reviewer: > Often requires composing multiple modules for complex configurations

This is true for Fabric too

   </td>
  </tr>
  <tr>
   <td><strong>Extensibility</strong>
   </td>
   <td>Built with extensibility in mind, catering to fork and use patterns. Modules are often lightweight and easy to adopt / tailor to specific use cases.
   </td>
   <td>Not primarily built for fork-and-use extensibility, catering primarily to off-the-shelf consumption. Modules are tailored towards common use cases and extensible via composition.
   </td>
  </tr>
  <tr>
   <td><strong>Config Customization</strong>
   </td>
   <td>Prefer customization using variables via objects, providing a tight but comprehensive variable space within modules.
   </td>

>    <td>Prefer customization using variables via primitives, which can lead to more modularity but may require using multiple modules for customization.

Reviewer: I don't see how more variables "can lead to more modularity but may require using multiple modules for customization"

Author: I was trying to convey that CFT seems to often prefer smaller, more focused modules, which *might* result in needing to use several of them together for complex configurations.  However, that's a design choice, not a direct consequence of variable count.

Should we remove that sentence altogether? Or perhaps rephrase it to focus solely on CFT's tendency towards smaller, more composable modules (if that makes sense), without mentioning variable count?

Reviewer: > Should we remove that sentence altogether? Or perhaps rephrase it to focus solely on CFT's tendency towards smaller, more composable modules (if that makes sense), without mentioning variable count?

To be honest, I don't know. I don't understand what you're tying to highlight. 

You mention "CFT seems to often prefer smaller, more focused modules". IMO this is not true either. To give you an example, let's compare GKE modules

CFT (excluding documentation and boilerplate):
```
❯ tokei terraform-google-kubernetes-engine/*.tf terraform-google-kubernetes-engine/modules
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 HCL                   148        29025        21324         3226         4475
 Pan                     1           18           18            0            0
 Shell                  14          611          302          213           96
-------------------------------------------------------------------------------
 Markdown               18         3667            0         3151          516
 |- HCL                 10          716          622           11           83
 |- Shell                1            5            5            0            0
 |- YAML                 1            8            6            2            0
 (Total)                           4396          633         3164          599
===============================================================================
 Total                 181        33321        21644         6590         5087
===============================================================================
```

Fabric:
```
❯ tokei gke-cluster-autopilot gke-cluster-standard
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 HCL                     8         1893         1650          143          100
-------------------------------------------------------------------------------
 Markdown                2          276            0          200           76
 |- HCL                  2          359          335           24            0
 (Total)                            635          335          224           76
===============================================================================
 Total                  10         2169         1650          343          176
===============================================================================
```

Looking just at HCL, it's 29 kloc vs 1.8 kloc. I find hard it to justify that being smaller and more focused.

Reviewer 3: Agreed with Julio. CFT is bloated, and prefers adding extra functionality via submodules which further complicates things.

   </td>
  </tr>
  <tr>
   <td><strong>Examples</strong>
   </td>
   <td>Thorough examples for individual modules and end-to-end examples composing multiple modules, covering a wide variety of use cases from foundations to solutions. 
   </td>
   <td>Examples for a module mostly focus on that individual module. Composition is often not shown in examples but in larger modules built using smaller modules.
   </td>
  </tr>
  <tr>
   <td><strong>Resources</strong>
   </td>

>    <td>Modules often encapsulate a broader set of resources, providing more functionality within a single unit.

Reviewer: "often" is ambiguous. Does it happen or not?

What are you trying to convey with this change?

Author: You're right, "often" is too vague. 

I was trying to highlight that Fabric modules sometimes handle a wider range of resources or functionality within a single module, compared to CFT's tendency to break things down into smaller, more focused units.

However, without concrete examples, this distinction becomes blurry.  

**Instead of a general statement, would it be more helpful to:**
1. **Remove the "Resources" row entirely:** If there's no clear and consistent difference in how Fabric and CFT handle resources *in general*, it might be best to remove this point of comparison. 
2. **Replace it with specific module comparisons:** We can use the Google Storage and Project modules as examples where Fabric consolidates more functionality. We could say something like:
> "Fabric modules, such as the `gcs` and `project` modules, often manage a broader set of resources within a single unit compared to CFT's more modular approach. For instance, Fabric's `gcs` module handles IAM, tag bindings, and lifecycle rules directly, while CFT might use separate modules or resources for those aspects."

What do you think would be the clearest way to represent this difference?

   </td>

>    <td>Heavier root modules often compose leaner sub-modules wrapping resources, which can lead to more granular control but potentially increased complexity.

Reviewer: what does "granular control" mean? What can you do with that control that Fabric doesn't let you do?

Author: Good point. "Granular control" needs to be more specific.

Rather than CFT inherently offering more control, it's about how that control is distributed. Fabric often provides a broader set of configuration options within a single module, while CFT might break those options into separate modules.

Take the same examples from our other discussion thread:

Google Storage: Fabric's `gcs` module handles IAM, tag bindings, etc., all in one place. CFT would require additional modules or resources for those aspects.
Project Setup: Fabric's `project` module covers IAM permissions, custom roles, and more. CFT's `project-factory` might handle some of those features through separate modules.

So, it's not necessarily about more control, but rather where that control is located. Fabric leans towards comprehensive modules, which can streamline configurations, while CFT's modularity might offer more flexibility in certain cases.

   </td>
  </tr>
  <tr>
   <td><strong>Resource Grouping</strong>
   </td>
   <td>Generally grouped by logical entities, aiming for a more holistic approach to resource management.
   </td>

>    <td>Generally grouped by products/product areas, aligning with Google Cloud's service organization. 

Reviewer: unnecessary clarification IMO

Author: Agreed. I'll remove that

```suggestion
   <td>Generally grouped by products/product areas.
```

  <tr>
   <td><strong>Factories</strong>
   </td>
   <td>Fabric implements several "factories" in modules, where users can drive or automate Terraform via YAML files (projects, subnetworks, firewalls, etc.). This simplifies common tasks and provides an alternative to pure Terraform configuration.
   </td>
   <td>CFT does not implement factories and generally shows examples usable with variable definitions files (.tfvars).
   </td>
  </tr>
   <tr>
   <td><strong>Variable Granularity</strong>
   </td>

>    <td>Offers a wider range of variables within a single module for greater flexibility. For example, a single FAST module might handle project creation with extensive customization options.

Reviewer: s/FAST module/Fabric module/

Author: ```suggestion
   <td>Offers a wider range of variables within a single module for greater flexibility. For example, a single Fabric module might handle project creation with extensive customization options.
```
